harness/determined/layers/_harness_profiler.py

```diff
@@ -180,3 +182,185 @@ def serialize_graph(self, path: str, figsize: Tuple[int, int] = (20, 40)) -> Non
         # before serializing to disk.
         plt.subplots_adjust(hspace=0.4)
         plt.savefig(path)
+
+
+class _ProfilingThread(threading.Thread):
+    """
+    Call a profiling function periodically on a separate thread (asynchronous collection).
+
+    The _ProfilingThread also supports synchronous triggering, which is useful for collecting
+    precise metrics around workload transitions.
+    """
+
+    def __init__(self, period: float, profile_fn: Callable) -> None:
+        self.period = period
+        self.profile_fn = profile_fn
+        self.quitting = False
+        self.cond = threading.Condition()
+        super().__init__()
+
+    def run(self) -> None:
+        """Asynchronously call a profiling function to collect metrics in a thread."""
+        with self.cond:
+            while True:
+                self.cond.wait(timeout=self.period)
+                if self.quitting:
+                    break
+                self.profile_fn()
+
+    def __enter__(self) -> None:
+        """
+        While in the context provided by the _ProfilingThread, the _ProfilingThread is guaranteed
+        not to trigger asynchronously.
+        """
+        self.cond.acquire()
+
+    def __exit__(self, *arg: Any) -> None:
+        self.cond.release()
+
+    def trigger(self) -> None:
+        """
+        Allow the profile_fn() to be triggered externally.  Only safe to call inside the context
+        manager provided by this self object.
+        """
+        self.profile_fn()
+
+    def quit(self) -> None:
+        self.quitting = True
```

Comment:
there's no need to; it would not affect the behavior of the system at all since it is a bool that is only changed once in the course of its lifetime. I have a tiny preference for leaving it out of the critical section since it doesn't literally have to be there.

But since there's a relevant critical section just below, I could move it there if you cared much at all.